### PR TITLE
Feature: Functionnality to add users by Xlsx.

### PR DESCRIPTION
**Pull request by Eyap53**
# PRE-ALPHA : Add users by Excel

### New Behaviors :

In this pre-alpha version, a users (the manager) that have the permission to add other users, can navigate by link to "/<group_name>/users/add_by_list_xlsx/".
This page can **ONLY** be accessed by a direct link, in order to limitate its use.

On this page, the manager can download the "template" for xlsx, containing the naming for each column.
He can also upload it once its completed. Borgia will analyse each line, and try to create the user. Be careful, if it fail, **IT WON'T BE DISPLAYED YET**. 
However, if the xlsx if correcty filled, it will add each user.

### Note :
This functionnality is released at this state by an urge demand by the campus of Angers, and will be improved in the future.

### CODE REVIEW

File: borgia/urls.py [hunk 1]
```diff
@@ -110,6 +110,14 @@
...
         UserSupplyMoney.as_view(), name='url_user_supplymoney'),
 
+        #       PRE-ALPHA : Ajout par xlsx
+    url(r'^(?P<group_name>[\w-]+)/users/add_by_list_xlsx/$',
+        AddUsersXlsx.as_view(),
+        name='url_add_users_xlsx'),
+    url(r'^(?P<group_name>[\w-]+)/users/add_by_list_xlsx_download/$',
+        AddUsersDownloadXlsx.as_view(),
+        name='url_add_users_download_xlsx'),
+
```
Comment (AlexandrePalo) [inline borgia/urls.py]: I do believe that the best may be :

/add_by_list/xlsx/
/add_by_list/xlsx/download

to open to further file type :)

File: users/templates/users/add_users_xlsx.html [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        <i class="fa fa-info-circle" aria-hidden="true"></i>
+        Informations
+    </div>
+    <div class="panel-body">
+      <p>Cette page permet d'ajouter des users par fichier excels.</p>
+      <p>La fonctionnalités est encore en mode PRE-ALPHA ! C'est a dire qu'elle n'as pas encore été testée totallement, et qu'elle contient potentiellement des bugs.</p>
+      <p>Pour cette raison, cette page n'est pas accessible depuis une navigation "normale" dans Borgia. Celle-ci sera integrée après un dévellopement plus poussé.</p>
+      <p>Dans tout les cas, Merci de faire <strong>VRAIMENT</strong> attention, cad de rentrer les utilisateurs petit à petit si possible, et de vérifier le résultat. MERCI de bien remplir les différents champs ! Par de caractères chelous, une adresse email valide, etc...
+      Si jamais une erreur se produit, elle n'est, pour l'instant, pas affichée. Vous ne pourrez donc pas savoir que tel ou tel utilisateurs n'a pas été rentré dans la base de donnée. Mais bon ça devrait quand même bien macher le travail.</p>
+    </div>
```
Comment (AlexandrePalo) [inline users/templates/users/add_users_xlsx.html]: Fautes d'orthographes à corriger hehe

File: users/views.py [hunk 3]
```diff
@@ -409,6 +411,86 @@ def get_initial(self):
...
         return initial
 
+
+class AddUsersXlsx(GroupPermissionMixin, FormView, GroupLateralMenuMixin):
+    """
+    Download/Upload Excel for adding users.
+    """
+    template_name = 'users/add_users_xlsx.html'
+    form_class = AddUsersUploadXlsxForm
+    perm_codename = 'add_user'
```
Comment (AlexandrePalo) [inline users/views.py]: Bonne idée d'utiliser cette perm :)

File: users/views.py [hunk 4]
```diff
@@ -409,6 +411,86 @@ def get_initial(self):
...
+                                           year=row[7].value)
+                user.set_password(form.cleaned_data['password'])
+                user.save()
+
+                # En deux fois au cas ou y'a un pb
+                user.groups.add(Group.objects.get(pk=5))
+                user.save()
+            except:
+                pass
+
```
Comment (AlexandrePalo) [inline users/views.py]: Peut être ajouter une colonne "membre d'honneur" ?
Pour vérifier l'appartenant au groupe Gadz.

File: users/views.py [hunk 5]
```diff
@@ -409,6 +411,86 @@ def get_initial(self):
...
+    def get(self, request, *args, **kwargs):
+
+        wb = Workbook()
+        # grab the active worksheet
+        ws = wb.active
+        ws.title = "Test"
+        ws.append(['Username (Ex: 53Me215)', 'Prenom', 'Nom','Email (un valide !)', 'Bucque (Ex: Eyap)', 'Fams (Ex: 53)',
+                        'Tabagns (Ex: Me)', 'Annee (Ex: 2015)', 'Mot de passe',
+                        'Note : Les champs sont obligatoire, sauf: bucque et nums. Pour rappel, cest une fonctionnalites en mode pre-alpha. Il ny aura pas de message derreur si ca plante. Faites attention. Vraiment.'])
+
```
Comment (AlexandrePalo) [inline users/views.py]: Fautes d'orthographes :p

File: borgia/urls.py (hunk 1 above)
Comment (Eyap53) [reply borgia/urls.py]: Good thinking, i will change it right away !

File: users/templates/users/add_users_xlsx.html (hunk 2 above)
Comment (Eyap53) [reply users/templates/users/add_users_xlsx.html]: oups ^^

**Comment by Eyap53** [comment]
Thanks ! 

Pour les mdp : Exact, je vais utiliser la génération, ça seras effectivement mieux. (J'avais gardé en esprit que la demande de réinitialisation était capricieuse, mais y'avait plus de soucis, on est d'accord ?)

Pour les "membres d'honneurs" : J'y avais pensé. Je pense que c'est mieux de pas le mettre, pour deux raisons : Déjà, il y en a normalement très peu, donc ils peuvent être ajoutés "à la main" (en le précisant dans la com). Ensuite, je voulais minimiser au maximum le nombre de colonne, pour éviter les mauvaises manipulations. Qu'est ce que tu en penses ?

Enfin, désolé pour les fautes d'orthographes, j'ai pas pris le temps de me relire, j'ai fait ça assez rapidement ^^'

**Comment by Eyap53** [comment]
Related issue : #36

**Comment by Eyap53** [comment]
Update : 
There is a urge need for the feature as of today (the new students want theirs accounts !).
I updated the branch with recent changes (including urls).

STILL NEEDED BEFORE ACCEPTING : 
- Tests for each view for get method.
- Tests for post method (main function of this feature).
- Maybe remove the pre-alpha warning ? <- Debatable

**Comment by rowsky** [comment]
> # Update
> There is a urge need for the feature as of today (the new students want theirs accounts !).
> I updated the branch with recent changes (including urls).
> 
> **Still needed before accepting this PR :**
> 
> * Tests for each view for get method.
> * Tests for post method (main function of this feature).
> * Maybe remove the pre-alpha warning ? <- Debatable

- I think that the upload feature is sufficiently stable to consider it more than a pre-alpha version. However nothing have been done yet for download. I will fix that by including at least the possibility to download a list of all users.
Later, it could be good to choose with user attribute to download, to filter users, to download by 50 or 100 or more ...

**Pull request closed without merging**